File: frontend/lib/controllers/order_controller.dart
import 'dart:convert';
import 'package:frontend/model/order_detail_model..dart' as detailModel;
import 'package:http/http.dart' as http;
import 'package:frontend/model/order_model.dart';

class OrderController {
  String baseUrl = 'http://localhost:3000/api/orders';
  // Get all orders
  Future<List<Order>> getOrders() async {
    try {
      final response = await http.get(Uri.parse(baseUrl));

      if (response.statusCode == 200) {
        List<dynamic> data = json.decode(response.body);
        return data.map((json) => Order.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load orders: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load orders: $e');
    }
  }

  // Get a specific order
  Future<Order> getOrder(String id) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/$id'));

      if (response.statusCode == 200) {
        Map<String, dynamic> data = json.decode(response.body);
        return Order.fromJson(data);
      } else {
        throw Exception('Failed to load order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load order: $e');
    }
  }

  // Modified to fetch order with details in a single request
  Future<Order> getOrderWithDetails(String id) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/$id?include_details=true'),
      );

      if (response.statusCode == 200) {
        Map<String, dynamic> data = json.decode(response.body);
        return Order.fromJson(data);
      } else {
        throw Exception(
          'Failed to load order with details: ${response.statusCode}',
        );
      }
    } catch (e) {
      throw Exception('Failed to load order with details: $e');
    }
  }

  // Create a new order
  Future<Order> createOrder(Order order) async {
    try {
      final response = await http.post(
        Uri.parse(baseUrl),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(order.toJson()),
      );

      if (response.statusCode == 201) {
        Map<String, dynamic> data = json.decode(response.body);
        return Order.fromJson(data);
      } else {
        throw Exception('Failed to create order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to create order: $e');
    }
  }

  // Update an existing order
  Future<Order> updateOrder(Order order) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/${order.id}'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(order.toJson()),
      );

      if (response.statusCode == 200) {
        Map<String, dynamic> data = json.decode(response.body);
        return Order.fromJson(data);
      } else {
        throw Exception('Failed to update order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to update order: $e');
    }
  }

  // Delete an order
  Future<bool> deleteOrder(String id) async {
    try {
      final response = await http.delete(Uri.parse('$baseUrl/$id'));

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to delete order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to delete order: $e');
    }
  }

  // Update the status of an order
  Future<Order> updateOrderStatus(String id, String status) async {
    try {
      final response = await http.patch(
        Uri.parse('$baseUrl/$id/status'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({'status': status}),
      );

      if (response.statusCode == 200) {
        Map<String, dynamic> data = json.decode(response.body);
        return Order.fromJson(data);
      } else {
        throw Exception(
          'Failed to update order status: ${response.statusCode}',
        );
      }
    } catch (e) {
      throw Exception('Failed to update order status: $e');
    }
  }

  // Get order details for a specific order
  Future<List<detailModel.OrderDetail>> getOrderDetails(String orderId) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/$orderId/details'));

      if (response.statusCode == 200) {
        List<dynamic> data = json.decode(response.body);
        return data
            .map((json) => detailModel.OrderDetail.fromJson(json))
            .toList();
      } else {
        throw Exception('Failed to load order details: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to load order details: $e');
    }
  }

  // Add order detail and update the order
  Future<Order> addOrderDetail(String orderId, OrderDetail detail) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/$orderId/details'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(detail.toJson()),
      );

      if (response.statusCode == 201) {
        final responseData = json.decode(response.body);
        // Return the updated order with the new detail
        return getOrderWithDetails(orderId);
      } else {
        throw Exception('Failed to add order detail: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to add order detail: $e');
    }
  }
}

File: backend_express/controllers/orderController.js
const OrderModel = require('../models/orderModel');

// Helper function to format order response
const formatOrderResponse = (order) => {
  if (!order) return null;
  
  return {
    id: order.id,
    customerId: order.customerId,
    orderDate: order.orderDate,
    totalAmount: order.totalAmount,
    status: order.status,
    customerName: order.customerName
  };
};

// Get all orders
exports.getAllOrders = async (req, res) => {
  try {
    const orders = await OrderModel.getAllOrders();
    res.status(200).json({
      success: true,
      data: orders.map(order => formatOrderResponse(order))
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to fetch orders',
      error: error.message
    });
  }
};

// Get a single order by ID
exports.getOrderById = async (req, res) => {
  try {
    let order;
    
    // Check if details should be included
    if (req.query.include_details === 'true') {
      order = await OrderModel.getOrderWithDetails(req.params.id);
    } else {
      order = await OrderModel.getOrderById(req.params.id);
    }
    
    if (!order) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }
    
    res.status(200).json({
      success: true,
      data: order
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to fetch order',
      error: error.message
    });
  }
};

// Create a new order
exports.createOrder = async (req, res) => {
  try {
    // Validate required fields
    const { customerId, orderDate, totalAmount } = req.body;
    
    if (!customerId || !orderDate || totalAmount === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: customerId, orderDate, totalAmount'
      });
    }
    
    const orderId = await OrderModel.createOrder(req.body);
    const newOrder = await OrderModel.getOrderById(orderId);
    
    res.status(201).json({
      success: true,
      message: 'Order created successfully',
      data: formatOrderResponse(newOrder)
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to create order',
      error: error.message
    });
  }
};

// Update an order
exports.updateOrder = async (req, res) => {
  try {
    // Validate required fields
    const { customerId, orderDate, totalAmount, status } = req.body;
    
    if (!customerId || !orderDate || totalAmount === undefined || !status) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: customerId, orderDate, totalAmount, status'
      });
    }
    
    const affected = await OrderModel.updateOrder(req.params.id, req.body);
    if (affected === 0) {
      return res.status(404).json({
        success: false,
        message: 'Order not found or no changes made'
      });
    }
    
    const updatedOrder = await OrderModel.getOrderById(req.params.id);
    
    res.status(200).json({
      success: true,
      message: 'Order updated successfully',
      data: formatOrderResponse(updatedOrder)
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to update order',
      error: error.message
    });
  }
};

// Update order status
exports.updateOrderStatus = async (req, res) => {
  try {
    const { status } = req.body;
    
    if (!status) {
      return res.status(400).json({
        success: false,
        message: 'Status is required'
      });
    }
    
    const affected = await OrderModel.updateOrderStatus(req.params.id, status);
    if (affected === 0) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }
    
    const updatedOrder = await OrderModel.getOrderById(req.params.id);
    
    res.status(200).json({
      success: true,
      message: 'Order status updated successfully',
      data: formatOrderResponse(updatedOrder)
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to update order status',
      error: error.message
    });
  }
};

// Delete an order
exports.deleteOrder = async (req, res) => {
  try {
    const affected = await OrderModel.deleteOrder(req.params.id);
    if (affected === 0) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }
    res.status(200).json({
      success: true,
      message: 'Order deleted successfully'
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to delete order',
      error: error.message
    });
  }
};

// Get order details
exports.getOrderDetails = async (req, res) => {
  try {
    const details = await OrderModel.getOrderDetails(req.params.id);
    res.status(200).json({
      success: true,
      data: details
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to fetch order details',
      error: error.message
    });
  }
};

// Add detail to order
exports.addOrderDetail = async (req, res) => {
  try {
    const { bookId, quantity, price } = req.body;
    
    if (!bookId || !quantity || price === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: bookId, quantity, price'
      });
    }
    
    const detailId = await OrderModel.addOrderDetail(req.params.id, req.body);
    
    // Get the updated order to reflect new total amount
    const updatedOrder = await OrderModel.getOrderById(req.params.id);
    
    res.status(201).json({
      success: true,
      message: 'Order detail added successfully',
      data: {
        detailId: detailId,
        order: formatOrderResponse(updatedOrder)
      }
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to add order detail',
      error: error.message
    });
  }
};

// Get order with details
exports.getOrderWithDetails = async (req, res) => {
  try {
    const order = await OrderModel.getOrderWithDetails(req.params.id);
    
    if (!order) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }
    
    res.status(200).json({
      success: true,
      data: order
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Failed to fetch order with details',
      error: error.message
    });
  }
};
